internal/regtest/matcher.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package regtest
+
+import (
+	"go.uber.org/yarpc/transport"
+
+	"github.com/golang/mock/gomock"
+)
+
+// Criterion is an argument that adds criteria to a transport request and
+// context matcher.
+type Criterion func(*matcher)
+
+// NewMatcher returns a pair of matchers corresponding to the arguments
+// of GetHandler(Context, Request), for verifying that GetHandler is called with
+// parameters that satisfy given constraints. Passing options like
+// WithService("foo") adds constraints to the matcher.
+func NewMatcher(criteria ...Criterion) (gomock.Matcher, gomock.Matcher) {
+	m := matcher{
+		constraints: make([]choiceConstraint, 0, 5),
+	}
+	for _, criterion := range criteria {
+		criterion(&m)
+	}
+	return gomock.Any(), &m
```

Comment:
On reflection of what I just wrote, I want to +1 on changing this to a generic RequestMatcher.  Then if we want to do special context matching we can create a ContextMatcher and combine them however we want.